lib/segment/src/index/field_index/binary_index.rs

```diff
@@ -0,0 +1,491 @@
+}
+
+/// Due to being able to store multi-values, the binary index is not a simple bitset, but rather a pair of bitsets, one for true values and one for false values.
+enum BinaryItem {
+    True,
+    False,
+    Both,
+    None,
+}
+
+impl BinaryMemory {
+    pub fn new() -> Self {
+        Self {
+            trues: BitVec::new(),
+            falses: BitVec::new(),
+        }
+    }
+
+    pub fn get(&self, id: PointOffsetType) -> BinaryItem {
+        debug_assert!(self.trues.len() == self.falses.len());
+        if (id as usize) < self.trues.len() {
+            unsafe {
+                // SAFETY: we just checked that the id is within bounds
+                match (
+                    self.trues.get_unchecked(id as usize).as_ref(),
+                    self.falses.get_unchecked(id as usize).as_ref(),
+                ) {
+                    (true, true) => BinaryItem::Both,
+                    (true, false) => BinaryItem::True,
+                    (false, true) => BinaryItem::False,
+                    (false, false) => BinaryItem::None,
+                }
+            }
+        } else {
+            BinaryItem::None
+        }
+    }
+
+    pub fn set_or_insert(&mut self, id: PointOffsetType, has_true: bool, has_false: bool) {
+        if !has_true && !has_false {
+            return;
+        }
+
+        if (id as usize) >= self.trues.len() {
+            self.trues.resize(id as usize + 1, false);
+            self.falses.resize(id as usize + 1, false);
+        }
+
+        debug_assert!(self.trues.len() == self.falses.len());
+
+        unsafe {
+            // SAFETY: we just resized the vectors to be at least as long as the id
+            self.trues.set_unchecked(id as usize, has_true);
+            self.falses.set_unchecked(id as usize, has_false);
+        }
+    }
+
+    /// Removes the point from the index and tries to shrink the vectors if possible. If the index is not within bounds, does nothing
+    pub fn remove(&mut self, id: PointOffsetType) {
+        if (id as usize) < self.trues.len() {
+            self.trues.set(id as usize, false);
+            self.falses.set(id as usize, false);
+        }
+
+        // TODO: should we avoid shrinking the vecs on every remove?
+        self.shrink();
+    }
+
+    /// Shrinks the vectors to the last populated index
+    fn shrink(&mut self) {
+        let last_populated_index = self.trues.last_one().max(self.falses.last_one());
+        match last_populated_index {
+            Some(index) if index < self.trues.len() - 1 => {
+                self.trues.truncate(index + 1);
+                self.falses.truncate(index + 1);
+            }
+            None => {
+                self.trues.clear();
+                self.falses.clear();
+            }
+            _ => {}
+        }
+    }
+
+    pub fn count_trues(&self) -> usize {
+        self.trues.count_ones()
+    }
+
+    pub fn count_falses(&self) -> usize {
+        self.falses.count_ones()
+    }
```

Comment:
Nice work on changing this to counting! I think counting in `set_or_insert` is not entirely correct though.

I'll append a commit to prevent going back and forth in comments here.

Reply:
The change was a lot smaller than I thought. Before my understanding was that `set_or_insert` also set bits back to `0`, but that isn't the case.

Anyway. Thanks for implementing this. Marking as resolved!